graspologic/embed/case.py

```diff
@@ -0,0 +1,200 @@
+                     to the leading eigenvector of the covariate matrix.
+            - float: Use a particular alpha-value.
+
+    assortative : bool, default = True
+        Embedding algorithm to use. An assortative network is any network where the
+        within-group probabilities are greater than the between-group probabilities.
+        Here, L is the regularized Laplacian, Y is the covariate matrix, and a is the
+        tuning parameter alpha:
+            - True: Embed ``L + a*Y@Y.T``. Better for assortative graphs.
+            - False: Embed ``L@L + a*Y@Y.T``. Better for non-assortative graphs.
+
+    n_components : int or None, default = None
+        Desired dimensionality of output data. If "full",
+        n_components must be <= min(Y.shape). Otherwise, n_components must be
+        < min(Y.shape). If None, then optimal dimensions will be chosen by
+        ``select_dimension`` using ``n_elbows`` argument.
+
+    n_elbows : int, optional, default: 2
+        If `n_components=None`, then compute the optimal embedding dimension using
+        `select_dimension`. Otherwise, ignored.
+
+    check_lcc : bool , optional, defult = True
+        Whether to check if input graph is connected. May result in non-optimal
+        results if the graph is unconnected. Not checking for connectedness may
+        result in faster computation.
+
+
+    References
+    ---------
+    .. [1] Binkiewicz, N., Vogelstein, J. T., & Rohe, K. (2017). Covariate-assisted
+    spectral clustering. Biometrika, 104(2), 361-377.
+    """
+
+    def __init__(
+        self,
+        alpha=None,
+        assortative=True,
+        n_components=None,
+        n_elbows=2,
+        check_lcc=False,
+    ):
+        super().__init__(
+            n_components=n_components,
+            n_elbows=n_elbows,
+            check_lcc=check_lcc,
+            concat=False,
+            algorithm="square",
+        )
+
+        if not isinstance(assortative, bool):
+            msg = "embedding_alg must be a boolean value."
+            raise ValueError(msg)
+        self.assortative = assortative
+
+        if not ((alpha is None) or isinstance(alpha, (float, int))):
+            msg = "alpha must be in {None, float, int}."
+            raise TypeError(msg)
+
+        self.alpha = alpha
+        self.latent_right_ = None
+        self.is_fitted_ = False
+
+    def fit(self, network, y=None):
+        """
+        Fit a CASE model to an input graph, along with its covariates. Depending on the
+        embedding algorithm, we embed
+
+        .. math:: L_ = LL + \alpha YY^T
+        .. math:: L_ = L + \alpha YY^T
+
+        where :math:`\alpha` is a tuning parameter which makes the leading eigenvalues
+        of the two summands the same. Here, :math:`L` is the regularized
+        graph Laplacian, and :math:`Y` is a matrix of covariates for each node.
+
+        Covariates are row-normalized to unit l2-norm.
```

Comment:
are they centered?

Reply:
No. They were for a bit, but didn't appear to make much of a difference at the time (this was an old implementation though, using tuning alg etc). Can add that in easily if you want, just one line of code

Reply:
is this mentioned in the paper at all? It is often something they say once and then assume the rest of the paper?

Reply:
All the paper says is 

"For continuous covariates, it can be beneficial to centre and scale the columns of X
before performing the analysis. The scaling helps satisfy condition (i) in Lemma 1, which ensures
that the top K eigenvectors of ¯L(α) contain block information."

The[ original R code which generated(?) fig. 1](https://github.com/norbertbin/CASC/tree/master/simulations) scales but doesn't center, although there's an option to center.

the ? is after generated because the parameters in that code were different than the parameters used in the paper, but I scoured Norbert Binkiewicz's github and that's all that was there. Tried emailing him but he didn't respond.

Could add centering back in? Or add `scaled` and `centered` as class parameters?

Reply:
Added `center_covariates` and `scale_covariates` as class parameters.
0b37867